src/Analysis/Ast/Impl/Types/PythonClassType.cs

```diff
@@ -372,40 +391,44 @@ public bool Equals(IPythonClassType other)
 
-            // Prevent reentrancy when resolving generic class where
-            // method may be returning instance of type of the same class.
-            if (!Push(classType)) {
-                return _processing.Value;
-            }
+            // Push/Pop should be lock protected.
+            lock (_lock) {
+                // Prevent reentrancy when resolving generic class where
+                // method may be returning instance of type of the same class.
+                if (!Push(classType)) {
+                    return _processing.Value;
+                }
 
-            try {
-                // Create specific bases since we may have generic types there.
-                // Match generic parameter names to base type parameter names.
-                // Consider 'class A(Generic[T], B[T], C[E]): ...'
-                var genericTypeBases = Bases.Except(genericClassParameters).OfType<IGenericType>().ToArray();
-                // Start with regular types, then add specific types for all generic types.
-                var bases = Bases.Except(genericTypeBases).Except(genericClassParameters).ToList();
-
-                // Create specific types for generic type bases
-                // it for generic types but not Generic[T, ...] itself.
-                foreach (var gt in genericTypeBases) {
-                    var st = gt.Parameters
-                        .Select(p => classType.GenericParameters.TryGetValue(p.Name, out var t) ? t : null)
-                        .Where(p => !p.IsUnknown())
-                        .ToArray();
-                    if (st.Length > 0) {
-                        var type = gt.CreateSpecificType(new ArgumentSet(st));
-                        if (!type.IsUnknown()) {
-                            bases.Add(type);
+                try {
+                    // Create specific bases since we may have generic types there.
+                    // Match generic parameter names to base type parameter names.
+                    // Consider 'class A(Generic[T], B[T], C[E]): ...'
+                    var genericTypeBases = Bases.Except(genericClassParameters).OfType<IGenericType>().ToArray();
+                    // Start with regular types, then add specific types for all generic types.
+                    var bases = Bases.Except(genericTypeBases).Except(genericClassParameters).ToList();
+
+                    // Create specific types for generic type bases
+                    // it for generic types but not Generic[T, ...] itself.
+                    foreach (var gt in genericTypeBases) {
+                        var st = gt.Parameters
+                            .Select(p => classType.GenericParameters.TryGetValue(p.Name, out var t) ? t : null)
+                            .Where(p => !p.IsUnknown())
+                            .ToArray();
+                        if (st.Length > 0) {
+                            var type = gt.CreateSpecificType(new ArgumentSet(st));
```

Comment:
Is this a recursion under lock?

Reply:
Yes as with other push/pops. However, specifically here it is always single task since we don't run analysis and access it at the same time. Actually I am wondering if locks here are needed at all...